### PR TITLE
Python 3 update for dropbox-cli

### PR DESCRIPTION
**Pull request by augiekim**

### CODE REVIEW

**Comment by raveit65** [comment]
I just testeded the PR, but the dropbox daemon doesn't start any more after downloading the proprietary dropbox code. It fails with
```
/usr/bin/dropbox:344: PyGIDeprecationWarning: GObject.idle_add is deprecated; use GLib.idle_add instead
  GObject.idle_add(self.on_done)
a bytes-like object is required, not 'str'
```
full start log (forgot the deprecation warnings)
```
[rave@mother ~]$ dropbox start -i
Starting Dropbox.../usr/bin/dropbox:303: PyGIDeprecationWarning: Since version 3.11, calling threads_init is no longer needed. See: https://wiki.gnome.org/PyGObject/Threading
  GObject.threads_init()
/usr/bin/dropbox:453: PyGTKDeprecationWarning: Stock items are deprecated. Please use: Gtk.Button.new_with_mnemonic(label)
  self.ok = ok = Gtk.Button(stock=Gtk.STOCK_OK)
/usr/bin/dropbox:458: PyGTKDeprecationWarning: Stock items are deprecated. Please use: Gtk.Button.new_with_mnemonic(label)
  cancel = Gtk.Button(stock=Gtk.STOCK_CANCEL)

(Downloading dropbox code)

/usr/bin/dropbox:334: PyGIDeprecationWarning: GObject.idle_add is deprecated; use GLib.idle_add instead
  GObject.idle_add(self.loop_callback, *ret)

(Error popup message)

/usr/bin/dropbox:344: PyGIDeprecationWarning: GObject.idle_add is deprecated; use GLib.idle_add instead
  GObject.idle_add(self.on_done)
a bytes-like object is required, not 'str'
/usr/bin/dropbox:341: PyGIDeprecationWarning: GObject.idle_add is deprecated; use GLib.idle_add instead
  GObject.idle_add(self.on_exception, e)
/usr/bin/dropbox:313: PyGTKDeprecationWarning: The keyword(s) "type, message_format" have been deprecated in favor of "message_type, text" respectively. See: https://wiki.gnome.org/PyGObject/InitializerDeprecations
  message_format = s)
/usr/bin/dropbox:313: PyGTKDeprecationWarning: The "flags" argument for dialog construction is deprecated. Please use initializer keywords: modal=True and/or destroy_with_parent=True. See: https://wiki.gnome.org/PyGObject/InitializerDeprecations
  message_format = s)
```
dropbox

**Comment by augiekim** [comment]
@raveit65, Thanks for reporting this. Could you please let me know the OS name and version you are running and python3-gobject version? I cannot repro it on Ubuntu 16.4.

**Comment by augiekim** [comment]
@raveit65 Also could you add `print(ex)` between line#387 and #388 to see what the exact failure is?
```
                def error(ex):
                    print(ex)
                    FatalVisibleError(ERROR_CONNECTING)
```

**Comment by augiekim** [comment]
@raveit65 I figured out where the bug is. I added a commit with the fix. Thanks for letting me know.

**Comment by raveit65** [comment]
> Could you please let me know the OS name and version you are running and python3-gobject version? I cannot repro it on Ubuntu 16.4.

It's fedora 29 and i added this PR on top of latest build from rpmfusion by me.
(nautilus-dropbox-2018.11.28-2.fc29)
http://pkgs.rpmfusion.org/cgit/nonfree/nautilus-dropbox.git/
http://koji.rpmfusion.org/koji/packageinfo?packageID=357

**Comment by raveit65** [comment]
Btw: Does rst2man.py run with python3, or do we need python2 here?

**Comment by augiekim** [comment]
> Btw: Does rst2man.py run with python3, or do we need python2 here?
> Edit:
> And for what is rst2man.py needed? :)

This PR is removing Python 2 runtime dependency. Probably we can remove build time dependency in another PR.
rst2man.py is used to generate Doxygen document and it still runs on Python 2.

File: dropbox.in [hunk 1]
```diff
@@ -299,19 +257,18 @@ class DownloadState(object):
...
-        db_path = DROPBOXD_PATH.encode(sys.getfilesystemencoding())
         f = open("/dev/null", "w")
         try:
-            a = subprocess.Popen([db_path, "/testrun", "0"], preexec_fn=os.setsid, cwd=os.path.expanduser("~"),
+            a = subprocess.Popen([DROPBOXD_PATH, "/testrun", "0"], preexec_fn=os.setsid, cwd=os.path.expanduser("~"),
                                  stderr=sys.stderr, stdout=f, close_fds=True)
-        except Exception, ex:
-            print ex
+        except Exception as e:
+            print (e)
```
Comment (samertm) [inline dropbox.in]: nit: no space between `print` and `(`

File: dropbox.in [hunk 2]
```diff
@@ -376,11 +333,11 @@ if GUI_AVAILABLE:
                         GObject.idle_add(self.loop_callback, *ret)
 
                         if self._stopped:
-                            thread.exit()
-                except Exception, ex:
-                    print ex
+                            _thread.exit()
+                except Exception as e:
+                    print (e)
```
Comment (samertm) [inline dropbox.in]: nit: no space between `print` and `(`

File: generate-deb.sh [hunk 3]
```diff
@@ -432,16 +432,16 @@ Source: dropbox
...
 
 Package: dropbox
 Replaces: nautilus-dropbox
 Breaks: nautilus-dropbox
 Provides: nautilus-dropbox
 Architecture: any
-Depends: procps, python-gi (>= 3.12), python (>= 2.7), \${python:Depends}, \${misc:Depends}, libatk1.0-0 (>= 2.10), libc6 (>= 2.19), libcairo2 (>= 1.13), libglib2.0-0 (>= 2.40), libgtk-3-0 (>= 3.10.8), libpango1.0-0 (>= 1.36.3), lsb-release, gir1.2-gdkpixbuf-2.0 (>= 2.30.7), gir1.2-glib-2.0 (>= 1.40.0), gir1.2-gtk-3.0 (>= 3.10.8), gir1.2-pango-1.0 (>= 1.36.3)
-Suggests: nautilus (>= 3.10.1), python-gpg (>= 1.8.0)
+Depends: procps, python3-gi (>= 3.12), python3 (>= 3.4.0), \${python:Depends}, \${misc:Depends}, libatk1.0-0 (>= 2.10), libc6 (>= 2.19), libcairo2 (>= 1.13), libglib2.0-0 (>= 2.40), libgtk-3-0 (>= 3.10.8), libpango1.0-0 (>= 1.36.3), lsb-release, gir1.2-gdkpixbuf-2.0 (>= 2.30.7), gir1.2-glib-2.0 (>= 1.40.0), gir1.2-gtk-3.0 (>= 3.10.8), gir1.2-pango-1.0 (>= 1.36.3)
+Suggests: nautilus (>= 3.10.1), python3-gpgme (>= 0.3)
```
Comment (samertm) [inline generate-deb.sh]: We want to suggest python3-gpg, which is newer than python3-gpgme (we got rid of python3-gpgme last year b/c it's obsolete). It's not showing up in ubuntu's package search right now b/c the index is broken for releases other than trusty. So this should be `python3-gpg (>= 1.8.0)`

Comment (augiekim) [reply generate-deb.sh]: Thanks! I was wondering why `python3-gpg` was not exposed on Ubuntu package list.

File: dropbox.in [hunk 4]
```diff
@@ -1083,20 +1038,20 @@ Prints out a public url for FILE (which must be in your public folder).
     try:
         with closing(DropboxCommand()) as dc:
             try:
-                console_print(dc.get_public_link(path=unicode_abspath(args[0].decode(sys.getfilesystemencoding()))).get(u'link', [u'No Link'])[0])
-            except DropboxCommand.CommandError, e:
-                console_print(u"Couldn't get public url: " + str(e))
-            except DropboxCommand.BadConnectionError, e:
-                console_print(u"Dropbox isn't responding!")
+                console_print(dc.get_public_link(path=os.path.abspath(args[0])).get(u'link', [u'No Link'])[0])
```
Comment (monsta) [inline dropbox.in]: Can `'link'` and `'No Link'` be used without `u` modifier now? I guess Python 3 makes it easier for all strings...

Comment (raveit65) [reply dropbox.in]: indeed, the `u` modifier was removed all over the commit.

File: generate-deb.sh [hunk 5]
```diff
@@ -432,16 +432,16 @@ Source: dropbox
 Section: gnome
 Priority: optional
 Maintainer: Dropbox <support@dropbox.com>
-Build-Depends: cdbs, debhelper (>= 9), build-essential, libnautilus-extension-dev (>= 3.10.1), libglib2.0-dev (>= 2.40), python-gi (>= 3.12), python-docutils
+Build-Depends: cdbs, debhelper (>= 9), build-essential, libnautilus-extension-dev (>= 3.10.1), libglib2.0-dev (>= 2.40), python-gi (>= 3.12), python-docutils, python3-gi (>= 3.12), python3-docutils
```
Comment (monsta) [inline generate-deb.sh]: Hmm, `python-gi` shouldn't be needed anymore... and `python3-docutils` isn't needed **yet** as `rst2man.py` is still Python 2.

File: dropbox.in (hunk 4 above)
Comment (augiekim) [reply dropbox.in]: Thanks @monsta I'll address this in a new PR ASAP!

File: generate-deb.sh (hunk 5 above)
Comment (augiekim) [reply generate-deb.sh]: Thanks @monsta Yes, it should be `python-docutils` for now.

Comment (augiekim) [reply generate-deb.sh]: Fixed by #69

File: dropbox.in (hunk 4 above)
Comment (augiekim) [reply dropbox.in]: Fixed by #69

**Pull request merged**